src/model/model.nit

```diff
@@ -1201,6 +1273,265 @@ abstract class MType
+			anchor: nullable MClassType): MType is abstract
+
+	redef fun can_resolve_for(mtype, anchor, mmodule)
+	do
+		if not need_anchor then return true
+		for t in operands do
+			if not t.can_resolve_for(mtype, anchor, mmodule) then
+				return false
+			end
+		end
+		return true
+	end
+
+	redef fun depth
+	do
+		var result = 0
+		for mtype in operands do
+			if mtype.depth > result then
+				result = mtype.depth
+			end
+		end
+		return result
+	end
+
+	redef var full_name is lazy do
+		var names = new Array[String].with_capacity(operands.length)
+		for mtype in operands do
+			names.add(mtype.full_name)
+		end
+		return names.join(separator)
+	end
+
+	redef fun is_legal_in(mmodule, anchor)
+	do
+		for mtype in operands do
+			if not mtype.is_legal_in(mmodule, anchor) then
+				return false
+			end
+		end
+		return true
+	end
+
+	redef fun is_ok
+	do
+		for mtype in operands do
+			if not mtype.is_ok then
+				return false
+			end
+		end
+		return true
+	end
+
+	redef var need_anchor is lazy do
+		for mtype in operands do
+			if mtype.need_anchor then
+				return true
+			end
+		end
+		return false
+	end
+
+	redef fun length
+	do
+		var result = 0
+		for mtype in operands do
+			result += mtype.length
+		end
+		return result
+	end
+
+	redef fun lookup_fixed(mmodule, resolved_receiver)
+	do
+		var resolved = new Set[MType]
+		for mtype in operands do
+			resolved.add(mtype.lookup_fixed(mmodule, resolved_receiver))
+		end
+		return apply_to(resolved, mmodule)
+	end
+
+	redef fun resolve_for(mtype, anchor, mmodule, cleanup_virtual)
+	do
+		var resolved = new Set[MType]
+		for t in operands do
+			resolved.add(t.resolve_for(mtype, anchor, mmodule, cleanup_virtual))
+		end
+		return apply_to(resolved, mmodule)
+	end
+
+	# The separator to use in `to_s` and `full_name`.
+	#
+	# `" and "` or `" or "`
+	protected fun separator: String is abstract
+
+	redef var to_s is lazy do return operands.join(separator)
+end
+
+# An intersection of multiple types.
+#
+# Conceptually, the subtypes that are common to all the `operands`.
+#
+# WARNING: Don’t instantiate this class directly: use `MType::intersection`
+# instead. Else, you may end up with undefined behaviors.
+class MIntersectionType
+	super MTypeSet[MType]
+
+	# Intersect the specified types.
+	init with_operands(mmodule: MModule, operands: MType...) do
+		init(mmodule, new HashSet[MType].from(operands))
+	end
+
+	redef fun ==(other)
+	do
+		# If we don’t redefine `==`, we would need to cache all intersections
+		# in the `Model` object (in order to make fixed-point algorithms work).
+		if super then return true
+		return other isa MIntersectionType and other.operands == operands
+	end
+
+	redef fun apply_to(operands, mmodule, anchor)
+	do
+		var i = operands.iterator
+		var result = i.item
+		i.next
+		for operand in i do
+			result = result.intersection(operand, mmodule, anchor)
+		end
+		return result
+	end
+
+	redef var as_notnull is lazy do
+		var new_operands = new Set[MType]
```

Comment:
More precisely, `T.as_not_null == T and Object`. However, because `Object` denies `null`, `intersection` would call `T.as_not_null` and… we get caught in an infinite recursion.